src/effects/native/echoeffect.cpp

```diff
@@ -95,88 +122,133 @@ void EchoEffect::processChannel(const ChannelHandle& handle, EchoGroupState* pGr
                                 const EffectProcessor::EnableState enableState,
                                 const GroupFeatureState& groupFeatures) {
     Q_UNUSED(handle);
-    Q_UNUSED(groupFeatures);
+
     DEBUG_ASSERT(0 == (numSamples % EchoGroupState::kChannelCount));
     EchoGroupState& gs = *pGroupState;
-    double delay_time = m_pDelayParameter->value();
+    // The minimum of the parameter is zero so the exact center of the knob is 1 beat.
+    double period = m_pDelayParameter->value();
     double send_amount = m_pSendParameter->value();
     double feedback_amount = m_pFeedbackParameter->value();
-    double pingpong_frac = m_pPingPongParameter->value();
-
-    int delay_samples = EchoGroupState::kChannelCount * delay_time * sampleRate;
-    VERIFY_OR_DEBUG_ASSERT(delay_samples <= gs.delay_buf.size()) {
-        delay_samples = gs.delay_buf.size();
+    double pingpong_fraction = m_pPingPongParameter->value();
+
+    int delay_length_samples;
+    if (groupFeatures.has_beat_length_sec) {
+        // period is a number of beats
+        if (m_pTripletParameter->toBool()) {
+           if (period <= 0.125) {
+              period = 0.041666666666666664; // 1/8 / 3
```

Comment:
Use her directly "1/8 / 3".  The preprocessor is able to calculate that at compile time.